<a href="https://colab.research.google.com/github/castrokelly/PPGIa/blob/main/atividade_3_pearson.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pontifícia Universidade Católica do Paraná — PUCPR

**Aluna:** 40131556 - Kelly Christine Alvarenga de Castro (christine.kelly@pucpr.edu.br)  
**Mestrado / Turma:** PPGIa (404) 2026/02  
**Disciplina:** Fundamentos de Matemática Computacional (Turma ME)  
**Professor:** Dr. Vinícius Mourão Alves de Souza

## ATIVIDADE 3
Normalização de vetores e Coeficiente de Correlação de Pearson

## Enunciado

1. Gerar um vetor **A** com 100 números reais aleatórios entre 0 e 100.
2. Gerar um vetor **B** com 100 números reais aleatórios entre 0 e 1.
3. Reescalar os dois vetores para o intervalo entre 5 e 10, usando uma função própria.
4. Calcular o **Coeficiente de Correlação de Pearson** entre os vetores reescalados usando apenas operações básicas do NumPy.
5. Comparar o resultado com `pearsonr` da biblioteca `scipy.stats`.
6. Calcular a correlação nos vetores originais, sem reescala, e comparar os resultados.

## 1. Bibliotecas e configuração

O **NumPy** será utilizado para criar os vetores e realizar as operações matemáticas. A função `pearsonr` do **SciPy** será usada somente no momento da comparação solicitada pelo exercício.

A semente aleatória é definida para que os mesmos vetores possam ser reproduzidos em novas execuções do notebook.

In [1]:
import numpy as np
from scipy.stats import pearsonr

np.random.seed(42)
np.set_printoptions(precision=4, suppress=True)

## 2. Geração dos vetores A e B

A função `np.random.rand(100)` gera 100 valores reais no intervalo $[0,1)$. Por isso:

- para **A**, multiplicamos os valores por 100, obtendo números no intervalo $[0,100)$;
- para **B**, usamos diretamente os valores produzidos por `np.random.rand(100)`, mantendo o intervalo $[0,1)$.

In [2]:
A = np.random.rand(100) * 100
B = np.random.rand(100)

print('Vetor A:')
print(A)
print()
print('Vetor B:')
print(B)

print()
print('Dimensão de A:', A.shape)
print('Dimensão de B:', B.shape)

Vetor A:
[37.454  95.0714 73.1994 59.8658 15.6019 15.5995  5.8084 86.6176 60.1115
 70.8073  2.0584 96.991  83.2443 21.2339 18.1825 18.3405 30.4242 52.4756
 43.1945 29.1229 61.1853 13.9494 29.2145 36.6362 45.607  78.5176 19.9674
 51.4234 59.2415  4.645  60.7545 17.0524  6.5052 94.8886 96.5632 80.8397
 30.4614  9.7672 68.4233 44.0152 12.2038 49.5177  3.4389 90.932  25.878
 66.2522 31.1711 52.0068 54.671  18.4854 96.9585 77.5133 93.9499 89.4827
 59.79   92.1874  8.8493 19.5983  4.5227 32.533  38.8677 27.1349 82.8738
 35.6753 28.0935 54.2696 14.0924 80.2197  7.4551 98.6887 77.2245 19.8716
  0.5522 81.5461 70.6857 72.9007 77.127   7.4045 35.8466 11.5869 86.3103
 62.3298 33.0898  6.3558 31.0982 32.5183 72.9606 63.7557 88.7213 47.2215
 11.9594 71.3245 76.0785 56.1277 77.0967 49.3796 52.2733 42.7541  2.5419
 10.7891]

Vetor B:
[0.0314 0.6364 0.3144 0.5086 0.9076 0.2493 0.4104 0.7556 0.2288 0.077
 0.2898 0.1612 0.9297 0.8081 0.6334 0.8715 0.8037 0.1866 0.8926 0.5393
 0.8074 0.8961 0.318  0.1101

## 3. Função própria para reescala Min-Max

Para transformar um valor $x$ do intervalo original para o novo intervalo $[5,10]$, será usado:

$$
x' = 5 + \frac{x - x_{min}}{x_{max} - x_{min}}(10-5)
$$

A função abaixo implementa essa transformação diretamente com operações do NumPy.

In [3]:
def rescale_vector(vector, new_min=5, new_max=10):
    old_min = np.min(vector)
    old_max = np.max(vector)

    scaled_vector = new_min + ((vector - old_min) / (old_max - old_min)) * (new_max - new_min)
    return scaled_vector

## 4. Reescala dos vetores para o intervalo [5, 10]

Mesma função aplicada aos dois vetores. Em seguida, verificamos os menores e maiores valores encontrados. Como a reescala usa o mínimo e o máximo observados em cada vetor, o menor valor passa a ser 5 e o maior passa a ser 10.

In [4]:
A_scaled = rescale_vector(A, 5, 10)
B_scaled = rescale_vector(B, 5, 10)

print('A reescalado:')
print(A_scaled)
print()
print('B reescalado:')
print(B_scaled)

print()
print('Intervalo de A reescalado:', np.min(A_scaled), 'a', np.max(A_scaled))
print('Intervalo de B reescalado:', np.min(B_scaled), 'a', np.max(B_scaled))

A reescalado:
[ 6.8801  9.8157  8.7013  8.022   5.7668  5.7666  5.2678  9.385   8.0345
  8.5795  5.0767  9.9135  9.2131  6.0537  5.8983  5.9063  6.522   7.6455
  7.1726  6.4557  8.0892  5.6826  6.4603  6.8385  7.2955  8.9723  5.9892
  7.5919  7.9902  5.2085  8.0673  5.8407  5.3033  9.8064  9.8917  9.0906
  6.5239  5.4695  8.458   7.2144  5.5936  7.4948  5.1471  9.6048  6.2903
  8.3474  6.56    7.6216  7.7573  5.9137  9.9118  8.9211  9.7586  9.531
  8.0181  9.6688  5.4227  5.9704  5.2023  6.6294  6.9522  6.3544  9.1942
  6.7895  6.4032  7.7369  5.6899  9.059   5.3517 10.      8.9064  5.9843
  5.      9.1266  8.5733  8.6861  8.9014  5.3491  6.7982  5.5622  9.3693
  8.1475  6.6578  5.2957  6.5563  6.6287  8.6892  8.2202  9.4922  7.3778
  5.5812  8.6058  8.848   7.8315  8.8999  7.4877  7.6352  7.1502  5.1014
  5.5216]

B reescalado:
[ 5.125   8.2158  6.5705  7.5627  9.6011  6.2381  7.0611  8.8245  6.1334
  5.3578  6.4448  5.7881  9.7141  9.093   8.2004  9.4166  9.0703  5.9176
  9.5244  7.7

## 5. Implementação do Coeficiente de Correlação de Pearson

Coeficiente da correlação de Pearson para dois vetores $x$ e $y$:

$$
r = \frac{\sum_{i=1}^{N}(x_i-\mu_x)(y_i-\mu_y)}
{\sqrt{\sum_{i=1}^{N}(x_i-\mu_x)^2\sum_{i=1}^{N}(y_i-\mu_y)^2}}
$$

Na implementação, busquei usar apenas cálculos elementares com NumPy: média, subtração, multiplicação, soma, potência e raiz quadrada. Não utilizei `np.corrcoef` nem outra função pronta de correlação.

In [5]:
def pearson_numpy(x, y):
    mean_x = np.mean(x)
    mean_y = np.mean(y)

    diff_x = x - mean_x
    diff_y = y - mean_y

    numerator = np.sum(diff_x * diff_y)
    denominator = np.sqrt(np.sum(diff_x ** 2) * np.sum(diff_y ** 2))

    r = numerator / denominator
    return r

## 6. Correlação entre os vetores reescalados

Cálculo  para correlação entre `A_scaled` e `B_scaled` com a função implementada acima.

In [6]:
r_scaled_numpy = pearson_numpy(A_scaled, B_scaled)

print('Correlação de Pearson - implementação com NumPy:')
print(r_scaled_numpy)

Correlação de Pearson - implementação com NumPy:
-0.03403250097882253


## 7. Comparação com `scipy.stats.pearsonr`

A função `pearsonr` fornece o coeficiente de correlação de Pearson e também um **p-value** associado ao teste de hipótese de correlação linear nula.

Para responder ao item 1 do exercício, o coeficiente retornado pelo SciPy é comparado com o valor obtido pela implementação anterior.

In [8]:
r_scaled_scipy, p_value = pearsonr(A_scaled, B_scaled)

difference = np.abs(r_scaled_numpy - r_scaled_scipy)

print('Minha implementação:', r_scaled_numpy)
print('scipy.stats.pearsonr:', r_scaled_scipy)
print('Diferença absoluta:', difference)
print('p-value:', p_value)

Minha implementação: -0.03403250097882253
scipy.stats.pearsonr: -0.034032500978822516
Diferença absoluta: 1.3877787807814457e-17
p-value: 0.7367625222563697


### Interpretação da comparação

Os dois coeficientes devem ser iguais, exceto por possíveis diferenças muito pequenas decorrentes da precisão numérica de ponto flutuante. Isso indica que a implementação direta da fórmula de Pearson está consistente com a função de referência do SciPy.

O `p-value` é apresentado apenas porque faz parte do retorno de `pearsonr`; o foco principal desta atividade é a comparação do **coeficiente $r$**.

## 8. Correlação nos vetores originais, sem reescala

Para responder ao item 2, calculamos novamente a Correlação de Pearson usando os vetores originais `A` e `B`. Depois comparamos esse valor com a correlação obtida após a reescala.

In [9]:
r_original = pearson_numpy(A, B)

difference_original_scaled = np.abs(r_original - r_scaled_numpy)

print('Correlação nos vetores originais:', r_original)
print('Correlação nos vetores reescalados:', r_scaled_numpy)
print('Diferença absoluta:', difference_original_scaled)

Correlação nos vetores originais: -0.03403250097882254
Correlação nos vetores reescalados: -0.03403250097882253
Diferença absoluta: 6.938893903907228e-18


## 9. Conclusão

Os resultados mostram que os vetores **A** e **B** apresentam uma correlação de Pearson muito próxima de zero.

Para os vetores reescalados para o intervalo $[5, 10]$, foi obtido:

$$
r \approx -0.03403
$$

Esse valor indica uma **correlação linear negativa extremamente fraca**, praticamente inexistente para a amostra gerada. Como os dois vetores foram produzidos a partir de valores aleatórios independentes, esse resultado é compatível com o esperado.

A implementação desenvolvida utilizando apenas operações básicas do NumPy produziu:

$$
r_{\text{NumPy}} = -0.03403250097882253
$$

Enquanto a função `pearsonr` do SciPy retornou:

$$
r_{\text{SciPy}} = -0.034032500978822516
$$

A diferença absoluta entre os dois resultados foi:

$$
1.3877787807814457 \times 10^{-17}
$$

Essa diferença é desprezível e decorre apenas da precisão numérica de ponto flutuante. Portanto, **a implementação própria reproduziu corretamente o resultado da função de referência do SciPy**.

Além disso, a função `pearsonr` retornou:

$$
p = 0.7367625222563697
$$

Como o valor de $p$ é elevado, não há evidência estatística, nesta amostra, para rejeitar a hipótese de ausência de correlação linear entre os vetores. Esse resultado é coerente com o coeficiente de Pearson muito próximo de zero.

Ao calcular a correlação utilizando os vetores originais, antes da reescala, foi obtido:

$$
r_{\text{original}} = -0.03403250097882254
$$

Para os vetores reescalados, foi obtido:

$$
r_{\text{reescalado}} = -0.03403250097882253
$$

A diferença absoluta entre os dois valores foi:

$$
6.938893903907228 \times 10^{-18}
$$

Portanto, **não existe diferença relevante entre a correlação calculada antes e depois da reescala**.

Isso ocorre porque a reescala Min-Max utilizada é uma transformação linear crescente da forma:

$$
x' = ax + b, \qquad a > 0
$$

Embora essa transformação altere os valores, o intervalo e a escala dos dados, ela preserva a relação linear entre as observações. No cálculo da Correlação de Pearson, os deslocamentos são eliminados pela centralização em relação à média e os fatores positivos de escala são cancelados pela normalização.

Assim, conclui-se que **a reescala foi útil para colocar os vetores A e B na mesma faixa de valores $[5, 10]$, mas não alterou a Correlação de Pearson entre eles**. Neste experimento, tanto antes quanto depois da normalização, os vetores apresentaram praticamente ausência de associação linear.

## Referências utilizadas na atividade

- SOUZA, V. M. A. **Fundamentals of Computational Mathematics — Week 03**. PUCPR, 2026.
- BROWNLEE, J. **Basics of Linear Algebra for Machine Learning**. Machine Learning Mastery, 2018. Seção sobre covariância e correlação.
- Material e estrutura de identificação da disciplina disponibilizados no projeto.